pandas/core/categorical.py

```diff
@@ -1364,6 +1365,54 @@ def sort_values(self, inplace=False, ascending=True, na_position='last'):
             return self._constructor(values=codes, categories=self.categories,
                                      ordered=self.ordered, fastpath=True)
 
+    def rank(self, method='average', na_option='keep',
+             ascending=True, pct=False):
+        """
+        Rank the values along a given axis.
+
+        Parameters
+        ----------
+        values : array-like
+            Array whose values will be ranked. The number of dimensions in this
+            array must not exceed 2.
+        method : {'average', 'min', 'max', 'first', 'dense'},
+            default 'average'
+            The method by which tiebreaks are broken during the ranking.
+        na_option : {'keep', 'top'}, default 'keep'
+            The method by which NaNs are placed in the ranking.
+            - ``keep``: rank each NaN value with a NaN ranking
+            - ``top``: replace each NaN with either +/- inf so that they
+                       there are ranked at the top
+            - ``bottom``: replace each NaN with either +/- inf so that they
+                       there are ranked at the bottom
+        ascending : boolean, default True
+            Whether or not the elements should be ranked in ascending order.
+        pct : boolean, default False
+            Whether or not to the display the returned rankings in integer form
+            (e.g. 1, 2, 3) or in percentile form (e.g. 0.333..., 0.666..., 1).
+        """
+        from pandas.core.series import Series
+        if na_option not in ['keep', 'top', 'bottom']:
+            raise ValueError('invalid na_position: {!r}'.format(na_option))
+
+        codes = self._codes.copy()
+        codes = codes.astype(float)
+        if self._ordered:
+            na_mask = (codes == -1)
+            codes[na_mask] = np.nan
+            codes = _ensure_float64(codes)
```

Comment:
instead of all this

at the top ``rank`` to the imports from ``pandas.core.algorithms``


```
if self.ordered:
   codes[codes==-1] = np.nan
return rank(codes, ......)
```